packages/valory/skills/mech_interact_abci/behaviours.py

```diff
@@ -0,0 +1,749 @@
+        self,
+    ) -> WaitableConditionType:
+        """Send Mech metadata to IPFS."""
+        metadata_hash = yield from self.send_to_ipfs(
+            self.metadata_filepath, self.metadata, filetype=SupportedFiletype.JSON
+        )
+        if metadata_hash is None:
+            return False
+
+        v1_file_hash = to_v1(metadata_hash)
+        cid_bytes = cast(bytes, multibase.decode(v1_file_hash))
+        multihash_bytes = multicodec.remove_prefix(cid_bytes)
+        v1_file_hash_hex = V1_HEX_PREFIX + multihash_bytes.hex()
+        ipfs_link = self.params.ipfs_address + v1_file_hash_hex
+        self.context.logger.info(f"Prompt uploaded: {ipfs_link}")
+        self._v1_hex_truncated = Ox + v1_file_hash_hex[9:]
+        return True
+
+    def _build_request_data(self) -> WaitableConditionType:
+        """Get the request tx data encoded."""
+
+        for request in self.synchronized_data.mech_requests:
+
+            # Skip the request if it has been sent already
+            if "tx_hash" in request:
+                continue
+
+            self._metadata = MechMetadata(prompt=request["prompt"], tool=request["tool"])
+
+            status = yield from self._mech_contract_interact(
+                "get_request_data",
+                "data", # calldata is saved in self.data
+                get_name(MechRequestBehaviour.request_data),
+                request_data=self._v1_hex_truncated,
+            )
+
+            batch = MultisendBatch(
+                to=self.collateral_token,
+                data=HexBytes(self.data),
+            )
+            self.multisend_batches.append(batch)
+
+        return True
+
+    def _get_price(self) -> WaitableConditionType:
+        """Get the price of the mech request."""
+        result = yield from self._mech_contract_interact(
+            "get_price", "price", get_name(MechRequestBehaviour.price)
+        )
+        return result
+
+    def _get_safe_tx_hash(self) -> Generator[None, None, bool]:
+        """Prepares and returns the safe tx hash."""
+        status = yield from self.contract_interact(
+            performative=ContractApiMessage.Performative.GET_RAW_TRANSACTION,  # type: ignore
+            contract_address=self.synchronized_data.safe_contract_address,
+            contract_public_id=GnosisSafeContract.contract_id,
+            contract_callable="get_raw_safe_transaction_hash",
+            to_address=self.params.mech_agent_address,
+            value=self.price,
+            data=self.request_data,
+            data_key="tx_hash",
+            placeholder=get_name(MechRequestBehaviour.safe_tx_hash),
+        )
+        return status
+
+    def _prepare_safe_tx(self) -> Generator[None, None, str]:
+        """Prepare the safe transaction for sending a request to mech and return the hex for the tx settlement skill."""
+        for step in (
+            self._send_metadata_to_ipfs,
+            self._build_request_data,  # FIXME: shouldnt we first build the metadata?
```

Comment:
Yes, this won't work. The metadata need to exist when sending them to IPFS in `_send_metadata_to_ipfs`, but here they are generated in the `_build_request_data` method.

You could call the send/build as many times as the mech requests in the synced data and remove the internal loop in the build method. Also `self._metadata = MechMetadata(prompt=request["prompt"], tool=request["tool"])` would need to be the first thing to perform.

Reply:
I check the trader repo and this is in the same order there. Why?

Reply:
Because it is for a single request and the metadata are already [prepared](https://github.com/valory-xyz/trader/blob/54d165b86aad59c1109029f8a72a2aeb71ffbef8/packages/valory/skills/decision_maker_abci/behaviours/decision_request.py#L124) in setup.